### PR TITLE
Add quaternions page

### LINKED ISSUE #38: New section: Quaternions
* [Quaternions.jl](https://github.com/JuliaGeometry/Quaternions.jl)
  * The most popular package that I am one of the maintainers.
  * [Discourse announcement](https://discourse.julialang.org/t/ann-quaternions-jl-v0-7-0/91368)
* [Quaternionic.jl](https://github.com/moble/Quaternionic.jl)
  * This package was created when Quaternions.jl was not well-maintained.
  * Actively maintained, and there seem different features than Quaternions.jl

Some packages define their own quaternions, but that might be replaced with `Quaternions.Quaternion` in the future (https://discourse.julialang.org/t/taking-quaternions-seriously/44834/70).

* Makie.jl should now use `Quaternions.Quaternion` ([Makie.jl/src/utilities/quaternions.jl](https://github.com/MakieOrg/Makie.jl/blob/f2970dcd77bc16f311f8bb3226ef7d716395b369/src/utilities/quaternions.jl#L1-L5)).
* There was an issue in ReferenceFrameRotations.jl. to be compatible with Quaternions.jl ([ReferenceFrameRotations.jl#25](https://github.com/JuliaSpace/ReferenceFrameRotations.jl/issues/25)).

Another package [Grassmann.jl](https://github.com/chakravala/Grassmann.jl) seems able to handle quaternions.

### PR DESCRIPTION
**Pull request by hyrodium**
This PR fixes #38.

(cc: @moble)

### CODE REVIEW

**Comment by hyrodium** [comment]
@KronosTheLate Do you have any thoughts on this PR? Should we wait for some other reviews?

**Comment by hyrodium** [comment]
@moble Hi, this PR adds a new comparison page for quaternions. I mentioned you because you are maintaining Quaternionic.jl. Feel free to comment if my comparison is misleading or incorrect. (sorry for lacking an explanation in the first comment!)

**Comment by moble** [comment]
Yep, I actually started a review when I saw your first comment but didn't finish it.  :)  Will finish it in the next hour or so.

**Comment by KronosTheLate** [comment]
I have to admit, I am in the middle of some quite intense exam preparations. But I do not expect development to stop only for my sake. If we can get you both as independent actors to "sign off" on this PR, I am all good with merging it.

The PR author has obviously approved it, so as soon as @moble gives the green light, we have two independant thumbs up, and @hyrodium is free to merge it.

File: pages/quaternions.md
Comment (moble) [inline pages/quaternions.md:32]: I wouldn't really say that `Quaternionic` is much different from `Base.Complex`; I followed the latter pretty closely *except* for when I generalized `T <: Real` (which was a relatively recent change).  Also, I think it's important that `Quaternionic` exports the abstract type and three concrete subtypes.  Plus, there's just a lot more stuff included, so I want to hit the highlights:

```suggestion
* This package exports `AbstractQuaternion` as well as three concrete subtypes: `Quaternion` for arbitrary quaternions, `Rotor` for quaternions with unit magnitude, and `QuatVec` for quaternions with zero scalar part (corresponding to ordinary three-vectors).  These allows specializations for faster and/or more precise results in those special cases.
* Each of the types parametrizes the type `T` of its components, as in`Quaternionic.Quaternion{T}`, but does *not* require `T <: Real`.  As a result, this package can handle [biquaternions](https://en.wikipedia.org/wiki/Biquaternion) for example.
* This package exports `imx`, `imy`, and `imz` (similar to `Base.im`), and their unicode counterparts `𝐢`, `𝐣`, and `𝐤`.
* Special care is taken to ensure that functions such as `log`, `exp`, `sqrt`, etc., are accurate and smooth near singularities and branch cuts, and to ensure that they are differentiable at those points.  In particular [`ChainRules`](https://github.com/JuliaDiff/ChainRules.jl) and [`ForwardDiff`](https://github.com/JuliaDiff/ForwardDiff.jl) are explicitly supported.
* Methods are directly included to permit transformation to and from various representations of rotations, such as Euler angles, spherical coordinates, axis-angle, and rotation-matrix representations.
* Several functions are included to find the optimal rotation to align two sets of points, or the optimal rotor to align two sets of rotors, and to measure distances between quaternions or rotors.
* This package also enables various ways of dealing with quaternion-valued functions of time, including
  * Linear interpolation, or `slerp`
  * Quadratic interpolation, or `squad`
  * Conversion to *and from* angular velocity as a function of time
  * Conversion to the "minimal-rotation" frame
* The arguments for `Quaternionic.Quaternion` is ordered in $w+xi+yj+zk$.
```

File: pages/quaternions.md
Comment (moble) [inline pages/quaternions.md:24]:
```suggestion
```

It didn't help that `Quaternions` wasn't actively maintained, but that's not really the reason `Quaternionic` exists.  They really are different packages, and I feel like this line implies that the more active maintenance now happening makes `Quaternionic` unnecessary, when really there's a lot more in `Quaternionic` than is in `Quaternions`.

**Review by moble — CHANGES_REQUESTED** [review]

File: pages/quaternions.md
Comment (hyrodium) [reply pages/quaternions.md:24]: I see. Then, what would be the most bad part about Quaternions.jl?

>In particular, [the most popular](https://github.com/JuliaGeometry/Quaternions.jl) of those has made certain design choices that conflict with my needs.

I thought this text in [Quaternionic.jl README](https://github.com/moble/Quaternionic.jl) was related to these issues:

* `Quaternions.Quaternion` had incomplete compatibility with `Base.Complex`. https://github.com/JuliaGeometry/Quaternions.jl/issues/62
* `Quaternions.Quaternion` had `norm::Bool` field. https://github.com/JuliaGeometry/Quaternions.jl/issues/60.

Comment (moble) [reply pages/quaternions.md:24]: Well, first I should say that I don't consider most of the differences "bad"; they're just incompatible with my needs.  Different people make different decisions about what should go in a package.  But:

1. I don't think the first issue you point out was exactly what I was worried about, but there are indeed several places where `Quaternions` assumes that ambiguity in the direction of the vector part should resolve in the `x` direction.  This conflicts with common usage in physics and the theory of special functions, where the `z` direction is the most natural choice.
2. The second issue you point out was certainly part of my biggest concern: the lack of type structure.  Having `Quaternion`, `Rotor`, and `QuatVec` really helps ensure that operations make sense and that methods can be implemented in the most accurate and efficient ways.

Also, I just generally felt that the `Quaternions` devs were pushing to keep it a small package, whereas there are a lot more things that I wanted quaternions to be able to do.

Comment (hyrodium) [reply pages/quaternions.md:24]: Thank you for the comment!

>Also, I just generally felt that the Quaternions devs were pushing to keep it a small package, whereas there are a lot more things that I wanted quaternions to be able to do.

I also have the same thoughts. I'm also a maintainer of [Rotations.jl](https://github.com/JuliaGeometry/Rotations.jl), and I thought keeping Quaternions.jl small would help other packages that depend on Quaternions.jl.

>the lack of type structure. Having `Quaternion`, `Rotor`, and `QuatVec` really helps ensure that operations make sense and that methods can be implemented in the most accurate and efficient ways.

There was a related discussion about having `Quaternions.AbstractQuaternion` (https://github.com/JuliaGeometry/Quaternions.jl/issues/79#issuecomment-1059742924). I was thinking that we don't need that because we don't have `Base.AbstractComplex` (https://github.com/JuliaLang/julia/issues/33246).

But I now feel it would be okay to have `Quaternions.AbstractQuaternion` because https://github.com/JuliaLang/julia/pull/35587 is almost ready to merge..

Comment (moble) [reply pages/quaternions.md:24]: Can you also remove the "This causes the following confusion." line, please?  After that, I'm satisfied.

**Review by moble — APPROVED** [review]
Looks good.  Thanks!

**Comment by KronosTheLate** [comment]
Awesome! Just skimming the page, I already learnt a think or two about quaternions and their state in the Julia package ecosystem. Thanks <3

**Pull request merged**
